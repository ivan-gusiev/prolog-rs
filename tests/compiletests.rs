extern crate insta;
extern crate parameterized;
extern crate prolog_rs;

#[cfg(test)]
mod compiletests {
    use insta::assert_display_snapshot;
    use parameterized::parameterized;
    use prolog_rs::{
        compile::{compile_program, compile_query},
        lang::parse_term,
        symbol::SymbolTable,
        util::{case, writeout_sym},
    };

    #[parameterized(input = {
        "p(Z, h(Z,W), f(W))",
        "D",
        "f(X, g(X,a))",
        "f(b, Y)",
        "a"
    })]
    fn test_query_compile(input: &str) {
        let mut symbol_table = SymbolTable::new();
        let query = parse_term(input, &mut symbol_table).unwrap();
        let instructions = compile_query(query);

        assert_display_snapshot!(case(input, writeout_sym(&instructions, &symbol_table)));
    }

    #[parameterized(input = {
        "p(Z, h(Z,W), f(W))",
        "D",
        "f(X, g(X,a))",
        "f(b, Y)",
        "p(f(X), h(Y, f(a)), Y)",
        "a"
    })]
    fn test_program_compile(input: &str) {
        let mut symbol_table = SymbolTable::new();
        let query = parse_term(input, &mut symbol_table).unwrap();
        let instructions = compile_program(query);

        assert_display_snapshot!(case(input, writeout_sym(&instructions, &symbol_table)));
    }
}
