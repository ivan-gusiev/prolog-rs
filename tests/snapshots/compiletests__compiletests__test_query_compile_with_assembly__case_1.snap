---
source: tests/compiletests.rs
expression: "case(input, collapse(result))"
---
f(X, g(X,a))
-----
000	put_structure a/0, X4
001	put_variable X3, X1	// X3=X
002	put_structure g/2, X2
003	set_value X3	// X3=X
004	set_value X4
005	call @10

